src/py42/clients/alertrules.py

```diff
@@ -0,0 +1,82 @@
+from py42.clients import BaseClient
+
+# Below comments are to discuss design and is not intended to be official documentation.
+
+
+class AlertRulesClient(BaseClient):
+    """
+    `add` and `remove_users` require ids, i.e ruleIds and userIds, irrespective of which
+    type of rules they belong to, whereas search is dependent on rule type.
+
+    {
+      "tenantId": "string",
+      "ruleId": "string",
+      "userList": [
+        {
+          "userIdFromAuthority": "string",
+          "userAliasList": [
+            "string"
+          ]
+        }
+      ]
+    }
+
+    {
+      "tenantId": "string",
+      "ruleId": "string",
+      "userIdList": [
+        "string"
+      ]
+    }
+
+    Search create & update have 3 different endpoints for each task, e.g for search
+        /api/v1/Rules/query-cloud-share-permissions-rule
+        /api/v1/Rules/query-endpoint-exfiltration-rule
+        /api/v1/Rules/query-file-type-mismatch-rule
+
+    1. Design it similar to detectionlists
+        Now based on future requirement to add/update rules etc,
+        We can have module AlertRules and below clients
+            CloudShare
+            EndpointExfiltration
+            FileTypeMismatch
+
+         e.g sdk.alertrules.cloudshare.search(),
+             sdk.alertrules.endpointexfiltration.search(),
+             sdk.alertrules.filetypemismatch.search(), similarly for add and update.
+
+             sdk.alertrules.add_users(), sdk.alertrules.remove_users(), etc
+
+    2. Create an AlertRulesClient focusing only on current requirement.
```

Comment:
I like this one better because I think it is similar to other clients in py42. A single client can use multiple endpoints if they are closely related. However, your first idea is pretty cool. Do we suspect there will be / could be other methods for the three different rule types? Do they have different IDs or is getting a rule by ID just get any rule type that has that ID?

Another idea might be combining some of the ideas behind other querying py42 does. For example, we can have a search method that takes a parameter that is either an object or a str that contains the endpoint str. That object or str could be one of several enum options.

For example

class CloudSharePermissionsRule(object):
  _endpoint = "query-cloud-share-permissions-rule"

  def __str__(): 
    return _endpoint

def search(rule_type):
  url = "{}/{}".format(base_uri, str(rule_type))
  request(url)
`

Reply:
I like the api for option 2 with the exception of `get`.

the thing I don't like about `get` is that if we add additional rule types, py42 will not automatically be searching them without an update, and that fact won't be obvious to the user.

I think I like @unparalleled-js 's idea here -- we can expose the types of rules as constants and require the type as a parameter to `get`, making it more clear that that is the only type of rule being searched.

Reply:
ok, makes sense. 

Also analysing Web UI, `console.us.code42.com`, 'manage alert' tab it fetches all Rules with below endpoint
   https://alert-service-east.us.code42.com/svc/api/v1/rules/query-rule-metadata, as opposed to what is mentioned in swagger document, https://fed-observer-default.core-int.cloud.code42.com/svc/swagger/index.html#/. 

We can choose to use the metadata endpoint instead of individual ones, then we don't have to specify the rule_type, we need to make a call on the same.

Meanwhile, I will start work with option 2 for add and remove users.

Reply:
makes sense to me!

Reply:
Sounds good!